sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,652 @@
+          logInfo(s"Read delta file for version $version of $this from $fileToRead")
+        }
+
+        rocksDbWriteInstance.commit(Some(getBackupPath(version)))
+        rocksDbWriteInstance.close()
+        rocksDbWriteInstance = null
+      } catch {
+        case e: IllegalStateException =>
+          logError(s"Exception while loading state ${e.getMessage}")
+          if (rocksDbWriteInstance != null) {
+            rocksDbWriteInstance.abort()
+            rocksDbWriteInstance.close()
+          }
+          throw e
+      }
+    }
+    logInfo(s"Loading state for $version takes $elapsedMs ms.")
+  }
+
+  private def loadSnapshotFile(version: Long): Boolean = {
+    val fileToRead = snapshotFile(baseDir, version)
+    val tmpLocDir: File = new File(getTempPath(version))
+    val tmpLocFile: File = new File(getTempPath(version).concat(".tar"))
+    try {
+      if (!fm.exists(fileToRead)) {
+        return false
+      }
+      logInfo(s"Will download $fileToRead at location ${tmpLocFile.toString()}")
+      if (downloadFile(fm, fileToRead, new Path(tmpLocFile.getAbsolutePath), sparkConf)) {
+        FileUtility.unTar(tmpLocFile)
+        if (!tmpLocDir.list().exists(_.endsWith(".sst"))) {
+          logWarning("Snaphot files are corrupted")
+          throw new IOException(
+            s"Error reading snapshot file $fileToRead of $this:" +
+              s" No SST files found")
+        }
+        FileUtils.moveDirectory(tmpLocDir, new File(rocksDbPath))
+        return true
+      }
+    } catch {
+      case e: Exception =>
+        logError(s"Exception while loading snapshot file $e")
+        throw e
+    } finally {
+      if (tmpLocFile.exists()) {
+        tmpLocFile.delete()
+      }
+      FileUtils.deleteDirectory(tmpLocDir)
+    }
+    return false
+  }
+
+  /** Optional method for providers to allow for background maintenance (e.g. compactions) */
+  override def doMaintenance(): Unit = {
+    try {
+      val (files: Seq[WALUtils.StoreFile], e1) = Utils.timeTakenMs(fetchFiles(fm, baseDir))
+      logDebug(s"fetchFiles() took $e1 ms.")
+      doSnapshot(files)
+      cleanup(files)
+      cleanRocksDBBackupInstances(files)
+    } catch {
+      case NonFatal(e) =>
+        logWarning(s"Error performing snapshot and cleaning up $this")
+    }
+  }
+
+  private def doSnapshot(files: Seq[WALUtils.StoreFile]): Unit = {
+    if (files.nonEmpty) {
+      val lastVersion = files.last.version
+      val deltaFilesForLastVersion =
+        filesForVersion(files, lastVersion).filter(_.isSnapshot == false)
+      if (deltaFilesForLastVersion.size > storeConf.minDeltasForSnapshot) {
+        val dbPath = getBackupPath(lastVersion)
+        val snapShotFileName = getTempPath(lastVersion).concat(".snapshot")
+        val f = new File(snapShotFileName)
+        f.delete() // delete any existing tarball
+        try {
+          val (_, t1) = Utils.timeTakenMs {
+            FileUtility.createTarFile(dbPath, snapShotFileName)
```

Comment:
Does the provider guarantee `dbPath` always contains the content of the version? HDFS state store provider always "loads" the version first and then snapshot it, so it guarantees snapshot is properly created.

Reply:
if `dbPath` exists, as per the code semantics, it will have backed up version. Every time we commit/close a version, a backed up version is created at a different folder location.

I will make changes to check if `dbPath` exists before creating TarFile

Creating backup helped in physical isolation between current active DB and prev versions but it comes with a cost. (Every micro-batch takes some extra time to complete). I will try to find any other alternative to save this cost.